### PR TITLE
Add support for Microsoft Terminology Service API closes #1653

### LINKED ISSUE #1653: Add support for Microsoft Terminology Service API
It would be great to have [Microsoft Terminology Service API][1] as an machine translation service in Weblate.

It doesn't provide full phrase translation, but extracts terminology, so it should be probably shown below full translations (thus reported with lower quality rank).

There is a Python client implementation in Pontoon:

https://github.com/mozilla/pontoon/blob/8a9b9fd465225e5c40c6ce95b9eb6e4dadb97a36/pontoon/machinery/views.py#L116-L164

[1]:https://www.microsoft.com/en-us/language/Microsoft-Terminology-API

### PR DESCRIPTION
**Pull request by Brunux**
Hi, this is my first contribution to the project the PR is not ready yet but I really like to receive some feedback in order to refactor some code so, I will make some comment on the code needed (also any other feedback would be appreciated :wink: )

Also, It would be great if someone could help me to get the test going, so far I been trying to run the test like this:
`./manage.py test weblate.trans.tests.test_machine.MachineTranslationTest.test_microsoft_terminology`
but unlucky it doesn't work.

I really like to improve this code :muscle:

Signed-off-by: Bruno Fosados bruno.fosados@gmail.com
Before submitting pull request, please ensure that:

- [x] Every commit has message which describes it
- [x] Every commit is needed on it's own, if you have just minor fixes to previous commits, you can squash them
- [ ] Any code changes come with testcase
- [ ] Any new functionality is covered by user documentation

### CODE REVIEW

File: weblate/trans/machine/microsoft.py [hunk 1]
```diff
@@ -178,3 +182,172 @@ def convert_language(self, language):
...
+                  </soapenv:Body>
+                </soapenv:Envelope>
+            """.format(xmlns=self.MS_TM_SOAP_XMLNS, soap_action=soap_action)
+        elif soap_action == 'GetTranslations':
+            source = kwargs.get('source', '')
+            language = kwargs.get('language', '')
+            text = kwargs.get('text', '')
+            max_result = 5
+            if soap_action and source and language and text:
+                payload = """
```
Comment (Brunux) [inline weblate/trans/machine/microsoft.py]: #1673  Using `from django.template.loader import get_template `
This next xml could be inside template and use the `get_template('machinery/microsoft_terminology.jinja')` in order to rendener the neede variables like `payload = template.render({'from_lang': 'en-us', 'to_lang': 'es-mx'...})`

File: weblate/trans/machine/microsoft.py [hunk 2]
```diff
@@ -178,3 +182,172 @@ def convert_language(self, language):
...
+            for translation in results:
+                translations.append(tuple([
+                    translation.find(xpath + soap_target_translated).text,
+                    int(translation.find(xpath + soap_target_confidence).text),
+                    self.name,
+                    translation.find(xpath + soap_target_original).text
+                ]))
+        return translations
+
+    def convert_language(self, language):
```
Comment (Brunux) [inline weblate/trans/machine/microsoft.py]: #1673 Terminology API only supports languages like `en-us` or `en-gb` but no simple notation like `en` or `es`.  So this brings an issue when using `en` or `es` (without localization) in an effort to solve this `convert_language()` search for a matching language with localization and returns the first it founds in the list of supported languages. as an example when using `en`  `convert_language()` will return `en-au`. Any comments?

Comment (nijel) [reply weblate/trans/machine/microsoft.py]: It's probably better to use list of default locales we already maintain for the expansion:

https://github.com/WeblateOrg/weblate/blob/12dbb38dc2fbbc4bf5f52bd0c55bf385f4f1908a/weblate/lang/data.py#L931-L1036

File: weblate/trans/machine/microsoft.py (hunk 1 above)
Comment (nijel) [reply weblate/trans/machine/microsoft.py]: Certainly using django templates would be better. The current code for example doesn't well handle special chars in the payload.

File: weblate/trans/machine/microsoft.py (hunk 2 above)
Comment (Brunux) [reply weblate/trans/machine/microsoft.py]: Awesome, will do

File: weblate/trans/machine/microsoft.py (hunk 1 above)
Comment (Brunux) [reply weblate/trans/machine/microsoft.py]: Great will do, @nijel is there a slack channel or gitter I have some issues running the test maybe someone can help me with that, or should I contact you directly through email? Thanks

Comment (nijel) [reply weblate/trans/machine/microsoft.py]: There is Gitter, but honestly it's empty most of the time, that's why it's not advertised anywhere - https://gitter.im/WeblateOrg/Lobby

**Comment by RobnIrene** [comment]
Spoofed

**Comment by Brunux** [comment]
@RobnIrene Sorry what you mean by that?

**Comment by nijel** [comment]
It seems there is something wrong with the tests:

```======================================================================
FAIL: test_microsoft_terminology (weblate.trans.tests.test_machine.MachineTranslationTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.13/lib/python2.7/site-packages/httpretty/core.py", line 1146, in wrapper
    return test(*args, **kw)
  File "/home/travis/build/WeblateOrg/weblate/weblate/trans/tests/test_machine.py", line 298, in test_microsoft_terminology
    self.assert_translate(machine)
  File "/home/travis/build/WeblateOrg/weblate/weblate/trans/tests/test_machine.py", line 149, in assert_translate
    self.assertTrue(len(translation) > 0)
AssertionError: False is not true
```

**Comment by Brunux** [comment]
Hi @nijel I'm sorry test are no good yet do you know if there is some documentation that I can be based on to fix them, I have been no able to really understand them, thanks men

**Comment by nijel** [comment]
Thanks for your contribution! I've merged your code and applied following cleanups on top of that:

* 4437aba5f Changelog entry for #1673 and #1653
* 96b95f96f Simplify request generation
* 62784a095 Use broader serach for terminology
* 3bae88f36 Fix tests for MicrosoftTerminologyService
* 2ab33e6b6 Avoid duplicating translate method
* b5d18a712 Rationalize variable usage
* eb76ef40a Generate response just once
* 8fdd04d87 Avoid configurable XMLNS in the templates
* e749950cc Remove unused methods from MicrosoftTerminologyService

**Pull request merged**